### PR TITLE
Errors and edge cases

### PR DESCRIPTION
**Pull request by brettz9**
- fix: typos in error messages
- test: add new Errors Fixture for checking of precise error messages on failure conditions
- test: complete `transform` and `errors` coverage
- test: completes `SymbolParslet`, `OptionalParslet`, `KeyValueParslet`, `predicateParslet`, `ImportParslet`, `ParameterListParslet`, and `VariadicParslet`
- test: check assertion errors
- test: import from index page to get complete coverage of those export lines
- test: cover `ObjectParslet`, `FunctionParslet`, `NamepathParslet`, catharsis and jpt transforms
- chore: update caniuse-lite as per warning

### CODE REVIEW

**Comment by brettz9** [comment]
If there is a way to simulate `EarlyEndOfParseError`, I'm not sure how, so I did that one manually. I am guessing that would be an unexpected type of error?

**Comment by simonseyock** [comment]
The exception is thrown when there is a complete expression followed by something unexpected. For example `name]` would throw that.

**Comment by brettz9** [comment]
Ah ok... I've amended to use your more contextualized example.

I've been chipping away at the edges here, but still haven't taken time to deeply understand the nature of the parsing algorithm itself. Still, hope this may be useful as I do think other than that, I've mostly employed real-world test cases.

**Comment by brettz9** [comment]
Ok, there was one more thing I could do--the assertion errors. If those can occur in code, feel free to delete the file or the relevant portions. Just figured it may save a little time if those were just guards.

**Comment by brettz9** [comment]
Ok, did all I think I can do--10 uncovered lines now. Any more I would do would need to either remove code or create contrived examples which shouldn't, as far as I can tell, occur in the normal course of the app.

**Comment by brettz9** [comment]
Not sure if you want the last commit but may help you identify dead code or such. Now just 3 uncovered lines.

**Comment by brettz9** [comment]
I've amended the last commit to include 100% coverage. But again, you may well not want some or any of that last commit especially (except for the part to enforce coverage), since they do not occur in the normal course of the app, but hopefully it demonstrates the only way I saw to replicate in case you do want to allow for the API to recover if the user is using the app in some custom way.

With all of this, hopefully we will be ready to close #57 !

File: test/api.errors.spec.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    it('Throws with non-`event` type `JsdocTypeSpecialNamePath` ', () => {
+      expect(() => {
+        const namePathGrammar = [
+          // Ensure this gets preference over name
+          createSpecialNamePathParslet({
+            allowedTypes: ['module'],
+
+            // Exclude its `createSpecialNamePathParslet` to behave like
+            //    `basePathGrammar` in the `pathGrammar` file
+            pathGrammar: pathGrammar.slice(0, -1)
```
Comment (simonseyock) [inline test/api.errors.spec.ts:36]: would be nicer to use filter on the name. Then these tests do not depend on the parslet position (see above)

File: test/api.errors.spec.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+          }),
+
+          // Exclude SpecialNamePath for which we already added our own
+          //   replacement above
+          ...pathGrammar.slice(0, -1)
```
Comment (simonseyock) [inline test/api.errors.spec.ts:41]: would be nicer to use filter on the name. Then these tests do not depend on the parslet position (see above)

File: test/api.errors.spec.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    it('Throws with non-expected inner type toward building `JsdocTypeSpecialNamePath` ', () => {
+      expect(() => {
+        const namePathGrammar = [
+          // Ensure this gets preference over name (for `null`)
+          specialTypesParslet,
+
+          // Exclude SpecialNamePath for which we already added our own
+          //   replacement above
+          ...pathGrammar.slice(0, -1)
```
Comment (simonseyock) [inline test/api.errors.spec.ts:72]: would be nicer to use filter on the name. Then these tests do not depend on the parslet position (see above)

File: test/fixtures/Fixture.ts [hunk 4]
```diff
@@ -50,6 +53,18 @@ type Results = {
...
+      expect(() => {
+        parse(fixture.input, mode)
+      }).to.throw(expectedErrorForMode)
+    })
+    return
+  }
+  if (fixture.modes === undefined) {
+    return
+  }
   if (fixture.modes.includes(mode)) {
```
Comment (simonseyock) [inline test/fixtures/Fixture.ts:68]:
```suggestion
  if (fixture.modes?.includes(mode)) {
```

**Review by simonseyock — COMMENTED** [review]
Nice work! Some comments, but I think I have to look at some things again.

File: test/FunctionParslet.spec.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+function parse (text: string): void {
+  // Replace other function parslet with one setting
+  //   `allowNamedParameters: undefined`
+  const grammar = [...jsdocGrammar]
+  grammar.splice(-10, 1, createFunctionParslet({
+    allowWithoutParenthesis: true,
+    allowNamedParameters: undefined,
+    allowNoReturnType: true
+  }))
```
[the remark this answers is not in the document]
Comment (brettz9) [reply test/FunctionParslet.spec.ts:14]: Yup, looks good and works locally.

File: test/ObjectParslet.api.spec.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    })
+
+    let i = 0
+    Parser.prototype.parseIntermediateType = (precedence: Precedence): IntermediateResult => {
+      if (i++ === 1) {
+        const ret: unknown = undefined
+        return ret as RootResult
+      }
+      return pt.call(parser, precedence)
+    }
```
[the remark this answers is not in the document]
Comment (brettz9) [reply test/ObjectParslet.api.spec.ts:43]: Late here now, but hope to resolve tomorrow.

[the remark this answers is not in the document]
Comment (brettz9) [reply test/ObjectParslet.api.spec.ts]: Sounds good--I added a commit to replace it with a sinon stub.

**Review by simonseyock — APPROVED** [review]
Looks good to me! Feel free to merge.
